### ISSUE TITLE
TypeError: Cannot read properties of undefined (reading 'withConfig')

### ISSUE THREAD

**Issue by manudeli** [styled-components/styled-components#3956]
### System:
 - OS: macOS 13.2
 - CPU: (10) arm64 Apple M1 Pro
 - Memory: 2.16 GB / 32.00 GB
 - Shell: 5.8.1 - /bin/zsh
### Binaries:
 - Node: 18.13.0 - ~/.nvm/versions/node/v18.13.0/bin/node
 - Yarn: 3.3.1 - ~/.nvm/versions/node/v18.13.0/bin/yarn
 - npm: 8.19.3 - ~/.nvm/versions/node/v18.13.0/bin/npm
### npmPackages:
 - styled-components: ^5.3.6 => 5.3.6 

## Reproduction
I'm making a ui library (including Flex, Stack, etc.) that wraps styled-components called @jsxcss/styled-components. This library works fine with create-react-app.

However, in the Next.js environment, `Server Error
It doesn't work well with the error TypeError: Cannot read properties of undefined (reading 'withConfig')`.

[This link](https://github.com/manudeli/jsxcss-styled-components-test) reproduces this error situation.

You can see the code of @jsxcss/styled-components at [this link](https://github.com/jsxcss/react/tree/main/packages/styled-components).

![image](https://user-images.githubusercontent.com/61593290/221966944-7b2ce5f1-f3b4-4471-b8d7-ac8a01aed360.png)

## Steps to reproduce
1. You can clone repository in [This link](https://github.com/manudeli/jsxcss-styled-components-test) reproduces this error situation.
2. npm install
3. npm run dev

## Expected Behavior
No server error.

## Actual Behavior
Error like that picture

**Comment by jtribble** [comment]
I ran into a very similar issue: theme is `undefined` but `ThemeProvider` is definitely being passed a non-empty theme.

In my case, the issue was that `ThemeProvider` was created in a separate package from where the theme was being referenced, and the packages were using **different versions of react-dom**!

```js
// package A, using react-dom@18
const App = () => (
  <ThemeProvider theme={theme}>
    <GlobalStyle />
    <p>hello</p>
  </ThemeProvider>
);

// package B, using react-dom@17 implicitly
const GlobalStyle = createGlobalStyle`
  body {
    font-family: ${props => props.theme.fontFamily};
  }
`
```

After explicitly adding `react-dom@18` to package B's package.json, the issue was magically fixed.

I'm guessing your issue is somewhat related. Maybe the places where you're setting and referencing the theme have different views of the DOM, or are using different react-dom versions?

**Comment by mryechkin** [comment]
@manudeli I see you're using Rollup. I just ran into the same issue with my library while upgrading from Rollup v2 to v3.

In my case I had to add the [`output.interop`](https://rollupjs.org/configuration-options/#output-interop) option and set it to `"auto"`:

```js
// rollup.config.js
export default {
  // ...
  output: {
    interop: 'auto',
  },
};
```

**Comment by Arcalise08** [comment]
> @manudeli I see you're using Rollup. I just ran into the same issue with my library while upgrading from Rollup v2 to v3.
> 
> In my case I had to add the [`output.interop`](https://rollupjs.org/configuration-options/#output-interop) option and set it to `"auto"`:
> 
> ```js
> // rollup.config.js
> export default {
>   // ...
>   output: {
>     interop: 'auto',
>   },
> };
> ```
> 
> They changed some defaults in v3 as outlined in the [migration guide](https://rollupjs.org/migration/#changed-defaults). By the way, `"compat"` also seems to work.

2 days of time wasted only to find this one setting fixed the issues with styled components immediately for nextjs libraries

**Comment by Dessader** [comment]
> @manudeli I see you're using Rollup. I just ran into the same issue with my library while upgrading from Rollup v2 to v3.
> 
> In my case I had to add the [`output.interop`](https://rollupjs.org/configuration-options/#output-interop) option and set it to `"auto"`:
> 
> ```js
> // rollup.config.js
> export default {
>   // ...
>   output: {
>     interop: 'auto',
>   },
> };
> ```
> 
> They changed some defaults in v3 as outlined in the [migration guide](https://rollupjs.org/migration/#changed-defaults). By the way, `"compat"` also seems to work.

I encountered this problem too, but after adding `interop: "auto"` to the bundle, the changes from `babel-plugin-styled-components`, which are necessary to support SSR, are no longer made. How did you solve this problem?

**Comment by luistak** [comment]
I'm having the same problem but only after i upgraded NextJs to 14+, my project is using a ui library built with vite but even after upgrading the vite version and changing the [`output.interop` option to `auto`](https://github.com/styled-components/styled-components/issues/3956#issuecomment-1492694219) it did not fix my problem :(

**Comment by AbdoIbrahim-artex** [comment]
I have this problem  but i do not use styled-components library

**Comment by Saggitarie** [comment]
I have a similar issue but not using nextjs
Do I need to set something inside vite.config.mts instead?

**Comment by consultor5ajunior** [comment]
> Does someone found a solution to this? changing interop does nothing here. I have that problem online in vites SSR build. Not in the website build.

same here! ;(

**Comment by leo1104** [comment]
use import statement as:

`import {styled} from styled-components; `

**Comment by lanwin** [comment]
> use import statement as:
> 
> `import {styled} from styled-components; `

But you dont use Typescript, right?

**Comment by arthursakemi** [comment]
I just solved this issue a second ago after some days of struggle.
My case:
Built a library with vite, that uses styled-components and a design system lib that also uses styled-components.

What solved for me:
in `vite.config.ts`
- set `lib.formats: ['es', 'umd']`
- set `rollupOptions.output.interop: 'auto'`

now in `package.json`:
```json
  "main": "./dist/lib.umd.cjs",
  "types": "./dist/lib.d.ts",
  "exports": {
    ".": {
      "import": "./dist/lib.umd.cjs",
      "require": "./dist/lib.umd.cjs"
    }
  }
```

**Comment by 297318720** [comment]
> use import statement as:
> 
> `import {styled} from styled-components; `

Your approach has indeed solved this problem

**Comment by jialiu54321** [comment]
I am using vite to build a library with the [library-mode](https://vitejs.dev/guide/build#library-mode) and seeing empty theme , similar to [above issue](https://github.com/styled-components/styled-components/issues/3956#issuecomment-1464453899). 
Dev mode works fine.
Tried all approaches mentioned above but no luck

**Comment by Rishikeshari7** [comment]
Uninstall unwanted styling npm packages if you install
For Example : i was using  npm i react-better-carsoule ,
when i uninstall this package then i was able to deploy my website

**Comment by bayasdev** [comment]
> I am using vite to build a library with the [library-mode](https://vitejs.dev/guide/build#library-mode) and seeing empty theme , similar to [above issue](https://github.com/styled-components/styled-components/issues/3956#issuecomment-1464453899). Dev mode works fine. Tried all approaches mentioned above but no luck

Same problem using rollup for building a library

**Comment by JohanGJensen** [comment]
I was able to solve it by importing my component dynamically and turning off ssr.

In my case:
My application where I was trying to import a vite built component library is using NextJS version 14 and pages router.

```
import dynamic from 'next/dynamic';

const Component = dynamic(async () => await import('imported-package'), {
  ssr: false,
});
```

This comes with the cost of rendering the component on the client instead of course.

**Comment by quantizor** [comment]
Closing — this issue is stale and likely resolved in newer releases. If still reproducible on the latest v6.x, please open a new issue.
